docs/markdown/Python/python/python-third-party-dependencies.md

```diff
@@ -533,6 +521,100 @@ the user:
 indexes.add = ["http://$USERNAME:$PASSWORD@my.custom.repo/index"]
 ```
 
+### Local requirements
+
+There are two ways to specify local requirements from the filesystem:
+
+* [PEP 440 direct references](https://www.python.org/dev/peps/pep-0440/#direct-references)
+
+```python 3rdparty/python
+python_requirement(
+    name="django",
+    # Use an absolute path to a .whl or sdist file.
+    requirements=["Django @ file:///Users/pantsbuild/prebuilt_wheels/django-3.1.1-py3-none-any.whl"],
+)
+
+# Reminder: we could also put this requirement string in requirements.txt and use the
+# `python_requirements` target generator.
+```
+
+* The option `[python-repos].find_links`
+
+```toml pants.toml
+[python-repos]
+# Use an absolute path to a directory containing `.whl` and/or sdist files.
+find_links = ["file:///Users/pantsbuild/prebuilt_wheels"]
+```
+```shell
+❯ ls /Users/pantsbuild/prebuilt_wheels
+ansicolors-1.1.8-py2.py3-none-any.whl
+django-3.1.1-py3-none-any.whl
+```
+```python 3rdparty/BUILD
+# Use normal requirement strings, i.e. without file paths.
+python_requirement(name="ansicolors", requirements=["ansicolors==1.1.8"])
+python_requirement(name="django", requirements=["django>=3.1,<3.2"])
+
+# Reminder: we could also put these requirement strings in requirements.txt and use the
+# `python_requirements` target generator
+```
+
+Unlike PEP 440 direct references, `[python-repos].find_links` allows you to use multiple artifacts
+for the same project name. For example, you can include multiple `.whl` and sdist files for the same
+project in the directory; if `[python-repos].indexes` is still set, then Pex/pip may use
+artifacts both from indexes like PyPI and from your local `--find-links`.
+
+Both approaches require using absolute paths, and the files must exist on your machine. This is
+usually fine when locally iterating and debugging. This approach also works well if your entire
+team can use the same fixed location. Otherwise, see the below section.
+
+#### Working around absolute paths
+
+If you need to share the lockfile on different machines, and you cannot use the same
```

Comment:
This seems like an abrupt transition that partially lost the plot above. What about direct references? It would be good to say there is no solution currently for those; so prefer find links is you need to have things work amongst diverse dev / CI setups.

Reply:
Direct references work with `--path-mappings`, as shown by the tests :) But I agree I don't explicitly address how to use them well enough. It's a bit weird: you want to keep the original absolute path from the lock creator in the `python_requirement`, but then use `--path-mappings` so the correct location gets used.

Reply:
They don't really work in any useful way if you can't use build root interpolation.

Reply:
Why? The tests don't use interpolation, they use a tmpdir in `/tmp`.

Reply:
Most uses by real people will need to be in-repo to be useful is my claim. That needs build root substitution to work. You can't do that with python requirements (today) in Pants. You can in Pex / Pip since they support env var interpolation in requirements files. You can in pants.toml. But you just can't in non-pants.toml Pants metadata today.

Reply:
To clarify, it works to do this: in a BUILD file or requirements text, put `ansicolors @ file:///Users/eric/colors.whl` and generate the lockfile. Then, consumers keep that same BUILD file, but set `[python-repos].path_mappings`. Pex will know how to load `ansicolors @ file:///Users/eric/colors.whl`. That's wonky and confusing to have `ansicolors @ file:///Users/eric/colors.whl` in the BUILD file with the path for only one person's machine, but it does work. 

Although, I think it does fall apart when another person goes to regenerate the lockfile. So...yeah, this is all so niche that we should probably only document `path_mappings` working with `find-links`. Brave souls can discover this weirdness with PEP 440 on their own.

Reply:
Yeah - I discount that wonky as not acceptable.

So, the other thing to think about here is local projects. Let's discard the idea of locking local projects outside the repo from consideration - so how do I lock a local project today? Pex did alot of feature work to support local project locks, but for a Pants user to use it effectively, FWICT they are stymied by lack of support for a portable way to say this project here inside the build root. Maybe there is a way to spell this that I'm missing using Pants. Using Pex you can say `this/relative/path` but I'm pretty sure Pants requirements parsing chokes on this.